### ISSUE TITLE
API proposal to preemptively determine audio thread overload (to avoid audio glitching)

### ISSUE THREAD

**Issue by JohnWeisz** [WebAudio/web-audio-api#1595]
The Web Audio API defines audio production apps as a supported use-case, such as wave editors, digital audio workstations, and the like. And it's indeed quite adequate at the task, we at [AudioNodes](https://audionodes.com) had great success implementing it for a remarkably wide variety of features (even despite its painful limitation of not working in WebWorker threads).

One common trait of these applications is the capability of accepting virtually unlimited user content, which will essentially always result in hitting the limit of audio processing capabilities on a given machine at one point -- i.e. there is so many audio content that the **audio thread simply cannot keep up with processing** everything, it becomes overloaded, no matter how well it is optimized.

I believe this is widely and well understood in the audio production industry, and usually the solution to help the user avoid overloading is displaying a warning indicator of some kind, letting the user know the audio processing thread is about to be overloaded, and audio glitching will occur (so the user can know they should go easy on adding more contents).

**Currently, the Web Audio API does not expose a straightforward API to facilitate monitoring audio processing load, or overload.**

It's possible to _figure it out_, mainly in special cases with above-web-standard privileges (such as an Electron app). However, this is quite difficult to get right (even from native C++ side) without implementations taking a spec-defined standard into consideration.

I'd wish to propose a small set of light, straightforward, low-privacy implication API additions to enable this:

```js
audioContext.isOverloaded(); // 'true' or 'false'
```

```js
audioContext.addEventListener("overloadchange", function (e) {
    e.isOverloaded; // 'true' or 'false'
});
```

For obvious reasons, this is an extension to the `AudioContext` interface, not `BaseAudioContext`, as offline rendering does not require overload detection for the mentioned reasons. Having an event dedicated for the same purpose avoids the need of a polling check.

It is up to implementations to decide how _exactly_ it is determined whether the audio thread is considered overloaded or not, optionally taking the AudioContext `latencyHint` setting into consideration.

This would enable Web Audio API-based apps let the user know about high audio thread load, and display possible options or hints at steps for the user to take to avoid audio glitching.

**Privacy implications**

Exposing this information should have little to no privacy implications, as (1) it is rarely clear why exactly the audio thread is overloaded (it could be due to low device capabilities, or high CPU use by other processes), and (2) it does not provide a more accurate way to determine device capabilities than what is already possible with a simple scripted benchmark.

**Comment by rtoy** [comment]
I think this is really hard to specify in any kind of meaningful way, let alone some kind of meaningful inter-operable way.

It's really hard to know if something is overloaded until after the fact, which is too late for your use case.  This is further complicated by the fact that the audio graph can be quite dynamic.  So you might have a nice graph working at 1% CPU.  Then you connect in a very complex subgraph suddenly that would want to use 200% CPU.  Is it ok to signal overloaded after the fact?

**Comment by hoch** [comment]
Here's something we can do:
A = The time duration of audio device callback interval.
B = The time spent on rendering a render quantum.
Then the UA can give users the ratio between B and A when queried. But the information should (or MUST?) be encoded into a coarse form to minimize the room for the fingerprint abuse. (Perhaps categories like very low, low, moderate, high, very high. Potentially with some random jitter on the polling timing.) Of course the heuristic and the transition between these categories need discussion as well.

I believe the metaphor on DAW's CPU bar is somewhat inappropriate, because the CPU gauge on GUI is not an exposed programmable API. What's being asked here is a JS API that can be abused programmatically for so many different reasons, so we should tread carefully.

In general, I support this idea if we can pass the security/privacy review.

**Comment by JohnWeisz** [comment]
@rtoy 

 > It's really hard to know if something is overloaded until after the fact, which is too late for your use case. This is further complicated by the fact that the audio graph can be quite dynamic.

Thinking about it some more, I think the following idea could be a (relatively) accurate approximation, although this is an implementation detail.

Since the audio rendering thread is a single thread, and there are platform specific APIs to determine which CPU core a given thread is running on, we compute the average load of this CPU core (e.g. over the last second or so) and determine whether it is above a certain threshold.

Say this is done on a one second poll check by default. When an `AudioNode` is joined into the processing graph, the polling is done immediately, or perhaps with a slight debounce. This can detect both steady and immediate changes in audio thread load.

-----

@hoch

 > But the information should (or MUST?) be encoded into a coarse form to minimize the room for the fingerprint abuse. (Perhaps categories like very low, low, moderate, high, very high. Potentially with some random jitter on the polling timing.) Of course the heuristic and the transition between these categories need discussion as well.

Interesting insights, this is primarily the reason I proposed a simple true/false overload detect (instead of a load percentage), which should provide a far smaller surface for fingerprinting.

**Comment by JohnWeisz** [comment]
@rtoy

Additionally, this -- by definition -- cannot always hard-protect against audio thread overloading and glitching, there are way too many factors for that (such as a heavy background process starting on the same machine, or the immediate addition of a large sub-graph as you mentioned).

This API would merely _help_ avoid glitching in many (perhaps most) common cases.

**Comment by rtoy** [comment]
I think a binary isOverloaded will almost never work because the overload point won't match the apps developers expectation.  A coarse CPU percentage that @hoch proposes at least allows the developer to decide what overload means.  

Which is not to say I support this idea. of overload :-)  I'd much rather not do this and let each browser integrate WebAudio into the browser's developer tools.  Then we can provide much more information including things like overload, but also how much cpu each node in the graph is using or mark out the heaviest subgraph or whatever else might be useful for analyzing your application.  And no spec change required which means these can be iterated much faster.

**Comment by JohnWeisz** [comment]
> I'd much rather not do this and let each browser integrate WebAudio into the browser's developer tools. Then we can provide much more information including things like overload, but also how much cpu each node in the graph is using or mark out the heaviest subgraph or whatever else might be useful for analyzing your application.

How this would solve the problem of audio thread overload detection on the end-user's machine?

The threshold of hitting max. load depends greatly on device capabilities and running "background" tasks, and is impossible to know in advance.

**Comment by rtoy** [comment]
Yeah, it doesn't really help. It helps you as a developer to know what's using up CPU so you can make intelligent trade-offs, but doesn't help you to know if a particular machine is having problems.

I just don't know what that metric would be and making it meaningful in all the different possibilities that cause audio glitches.

**Comment by andrewmacp** [comment]
I agree that it would be really valuable to have some kind of API to expose this information, whether it's @hoch's coarse CPU use or @JohnWeisz's callback. After-the-fact could even be fine I think for our use cases. Better integration with dev tools would also be really great but an API would in addition allow dynamically reconfiguring/modifying the graph as we approach the resource limits on a given client.

**Comment by JohnWeisz** [comment]
To clarify, our use case would be simply displaying a warning notification about the user approaching maximum audio thread capacity -- as mentioned before, it's a DAW with no artificial limitation on the amount of content, and it would be great if the user was given an indication about audio overload, _other than glitching out audio_.

**Comment by rtoy** [comment]
Here is an alternative that you can implement now without waiting for the spec to get updated.

Create an offline context with an appropriately heavy graph.  Time how long it takes to render.  There's your CPU usage meter. You could probably do this for each DAW component to get an idea of the expense of adding the component.  You won't get the feedback that built-in cpu meter would give, but perhaps good enough.

**Comment by karlt** [comment]
One thing to bear in mind with such an API is that it is generally the rendering quantum that takes the most cpu that will be most likely to cause a break in the output.  I'd imagine an API similar to the one proposed by hoch in https://github.com/WebAudio/web-audio-api/issues/1595#issuecomment-386123070 but reporting a maximum ratio over some time period.

**Comment by JohnWeisz** [comment]
@rtoy

 > Here is an alternative [...] Create an offline context with an appropriately heavy graph. Time how long it takes to render. There's your CPU usage meter.

The problem with this approach is that it lags behind significantly, and rapidly allocating OfflineAudioContext has a significant performance overhead. Worse, it will run on a different thread than the real-time audio thread (i.e. AudioContext), and will merely give a generic hint at what the CPU usage can be.

Currently, the easiest but still relatively accurate approach is monitoring the load of each CPU core (this requires above-web-standard privileges, such as Electron), and considering the highest load as _the_ CPU load. This often results in false positives, but if the audio thread is indeed approaching overload, it will detect it.

**Comment by Mr21** [comment]
@JohnWeisz what is the priority for your application, live or rendering?

If it's rendering, we could ask for an audioContext method to modify in live the `BaseAudioContext.sampleRate`, having this possibility could be great, the user will choose in live when to downgrade the audio to avoid glitchs.

No need to have a measure. And maybe this can be done.

**Comment by rtoy** [comment]
Do you mean changing the sample rate of the context after it's been created?  Switching the sample rate can cause glitches from the switch, and you still have to resample the audio if the hardware doesn't support the new rate.  The resampling introduces additional delay and CPU usage.  That might not be what the user wants, and it's kind of hard to know what the additional delay and CPU would be.

**Comment by Mr21** [comment]
@rtoy, ah... and is there no way to reduce the context quality after created? I was thinking that reducing the sampleRate was something like reducing the resolution of a video.

**Comment by rtoy** [comment]
Currently no, there's no  way.  And yes, reducing the sample rate is kind of like reducing the video resolution.  But we still have to resample to the HW rate, and it's not clear what the right tradeoff should be.  High quality output (with high CPU usage and possibly significant additional latency)?  Low quality output (with lower CPU usage and possibly less additional latency)?  Something in between?

It's hard to know what the user wants when just switching sample rates.  This is one reason Chrome has not yet implemented the AudioContext sampleRate option; we don't know what the right tradeoff should be, but Chrome generally prefers high quality when resampling is needed.

**Comment by JohnWeisz** [comment]
@Mr21 It's both live _and_ rendering (you can play your project in real time, or render to an audio file), but AFAIK, rendering cannot glitch ever, as it will simply render slower if available CPU power is insufficient. This is only applicable to live playback, where there is only a limited time-frame to process a render quantum.

**Comment by Mr21** [comment]
@JohnWeisz, yes it's what i've though.
@rtoy, understood. In general, does the hardware rate can be modified (with the drivers or something) or it's clearly impossible?

**Comment by JohnWeisz** [comment]
In case the binary `isOverloading` reading is deemed insufficient and an exact CPU-usage reading having too severe privacy implications, Sherlock here has an alternative idea:

```js
audioContext.getOverloadState(); // "none" | "moderate" | "critical"
```

```js
audioContext.addEventListener("overloadchange", (e) => {
    e.overloadState; // "none" | "moderate" | "critical"
});
```

Where `moderate` means a considerably used audio thread, and `critical` means glitching is very likely.

This:

- provides a 2-step overload detection API for apps, so

**Comment by rtoy** [comment]
I like @hoch's idea:  Just return a CPU percentage, rounded to say the nearest 10% ro 20% or something coarse, but not too coarse.  Then we don't have to make random assumptions on the meaning of "none", "moderate", or "critical".

**Comment by hoch** [comment]
> How exactly this is determined is still up for implementation

So now developers need to handle unpredictable difference between devices/browser. The sniffing part will be really ugly. I did not use the term "CPU percentage" for this specific reason. As I suggested above, the ratio between the render timing budget and the time spent on rendering is quite clear and there's no room for creative interpretation.

IMO the sensible form for this attribute is:
```webidl
enum RenderPerformanceCategory {
  "none",
  "moderate",
  "critical"
}

partial interface AudioContext : BaseAudioContext {
  // renderPerformance = timeSpentOnRenderQuantum / callbackInterval
  readonly attribute RenderPerformanceCategory renderPerformance;
}
```

**Concerns**

This technically gives you an estimate of the performance of real-time priority thread provided by OS. That is why I believe the more coarse the better. Even with 3 categories, you can infer 3 classes of CPU powers from all of visitors. This is why the web platform does not have any API like this.

```js
const context = new AudioContext();
const mutedGain = new GainNode(context, { gain: 0.0 });
mutedGain.connect(context.destination);
let oscCounter = 0;

function harassAndCheck() {
  const osc = new OscillatorNode(context);
  osc.connect(mutedGain);
  osc.start();
  oscCounter++;
  if (context.renderPerformance !== 'critical') {
    setTimeout(AbuseAndCheck, 10);
  } else {
    console.log('I can run ' + oscCounter + ' oscillators!');
  }
}

harassAndCheck();
```

If you want to have a more fine-grained estimation, gain nodes can be exploited instead of oscillators. If you want to detect it fast, go with convolvers. So the coarseness of categories can be extrapolated by using various features in Web Audio API.

Here I am not trying to discourage the effort or anything, but simply want to check all the boxes until we reach the conclusion. Personally I want to have this feature in the API.

**Comment by JohnWeisz** [comment]
@hoch 

 > This technically gives you an estimate of the performance of real-time priority thread provided by OS.

I believe the main question here is whether the snippet you demonstrated can expose more information than what is already available.

I believe `AudioWorklet` can be used to achieve a rather accurate measure: AFAIK, the processing script is executed on the exact same real-time priority audio thread. I think this is significantly more accurate than what you could get with the `getOverloadState()` or `renderPerformance` API.

**Comment by hoch** [comment]
> I believe AudioWorklet can be used to achieve a rather accurate measure

Can you elaborate? I am curious about what can be done with the AudioWorklet. If that's possible, we can actually build a prototype with the AW first. On the other hand, the AW might have some privacy concerns we have not explored.

> AFAIK, the processing script is executed on the exact same real-time priority audio thread.

This is not true for Chrome; the AudioWorklet requires the JS engine on the rendering thread, so it actually degrades the thread priority when it is activated by `.addModule()` call. I am quite certain that it will be similar to all the browsers.

**Comment by hoch** [comment]
My team mates suggested to involve the privacy expert in the early stage of discussion and to review some cases like [this](https://eprint.iacr.org/2015/616.pdf).

@lknik Could you take a look at this? Would love to have your opinion here.

**Comment by Mr21** [comment]
ahhh... indeed this could increase our fingerprint on the web by measuring the audiocard performance :/
But anyway, how a PWA could not be able (technically) to do a fingerprint by benchmarks a random user?

**Comment by JohnWeisz** [comment]
@Mr21 
The primary privacy concern here is that this proposal would allow fingerprinting with a considerably greater accuracy, as you are indirectly benchmarking with the help of a real-time priority thread. This is much less likely to be affected by other running programs on the user's system, and your fingerprint will be potentially more accurate than a simple scripted benchmark.

**Comment by rtoy** [comment]
AudioWorklets should not run on a real-time priority thread. Chrome's worklet definitely doesn't.

**Comment by JohnWeisz** [comment]
> I am curious about what can be done with the AudioWorklet. If that's possible, we can actually build a prototype with the AW first.

I don't quite expect AudioWorklet to be *that* useful for determining CPU load in practice (perhaps the best you can do is measure how much looping you need before glitching out), but I was under the impression it is useful to benchmark the hardware itself (in this case, the `renderPerformance` API wouldn't introduce a _new_ fingerprint surface).

 > This is not true for Chrome; the AudioWorklet requires the JS engine on the rendering thread, so it actually degrades the thread priority when it is activated by .addModule() call. I am quite certain that it will be similar to all the browsers.

If I understand right, [this](https://bugs.chromium.org/p/chromium/issues/detail?id=813825) seems to be relevant (469639 referenced within is non-public, so I can only guess).

Then I was indeed mistaken (but ouch, this also means an audio worklet processor has a permanent potential stability impact _just by being loaded?_). In this case, discard my previous assumptions, AudioWorklet is clearly not useful for benchmarking a _real-time_ priority thread (according to the linked issue, only a "display-priority" one), and might not be considerably better than a simple scripted benchmark on the main thread. I should've studied the implementation more in depth.

**Comment by hoch** [comment]
> in this case, the renderPerformance API wouldn't introduce a new fingerprint surface

Except that it does expose something new! If the [battery status API](https://eprint.iacr.org/2015/616.pdf) can be used as a vector for side-channel attack, the `renderPerformance` API can certainly be worse if not careful. Perhaps throttling the update frequency would be good enough, but I would like to have opinions from privacy/security experts here.

> 469639 referenced within is non-public, so I can only guess

crbug.com/469639 is a launch tracking bug for the AudioWorklet. I am not sure why it's restricted. I'll talk to the team to unlock it.

**Comment by JohnWeisz** [comment]
> Except that it does expose something new! If the battery status API can be used as a vector for side-channel attack, the renderPerformance API can certainly be worse if not careful.

I believe you misunderstood my reply there (that, or I'm missing something). What I'm saying is that if the AudioWorklet _could_ be used to accurately benchmark a real-time thread (it cannot be, according to your reply), _then_ `renderPerformance` wouldn't expose anything new in practice (assuming you can do something like deliberately glitching out an audio worklet processor _and detecting the glitch_, so this assumption might be false).

As things stand, `renderPerformance` _does_ expose a new surface for fingerprinting, at least without severe adjustments to its return value.

 > Perhaps throttling the update frequency would be good enough, but I would like to have opinions from privacy/security experts here.

Wouldn't this just make building a fingerprint take longer? Assuming the throttle rate is known (which it _is_ in practice, if it's constant), just wait a bit between adding new nodes to the graph.

**Comment by hoch** [comment]
IIUC, fingerprinting or side-channel attack does not require the real-time thread. Certainly you can extract more info if the thread has the highest priority in the system, though. Fingerprinting is actually less of an issue here since Web Audio API has been [an abundant source of fingerprinting](https://audiofingerprint.openwpm.com/) since its beginning.

> Wouldn't this just make building a fingerprint take longer? 

Sorry, I was not clear. Throttling is a mitigation for the high precision timer + side-channel attack.

I guess we shared a fair amount of speculation so far, so I'll wait for the verdict from the security folks.

**Comment by hoch** [comment]
Another reference: [Chrome extension API for process inspection](https://developer.chrome.com/extensions/processes#type-Process).

**Comment by JohnWeisz** [comment]
Was there any progress on this one recently?

**Comment by hoch** [comment]
We have not gotten any feedback from the privacy folks. Not sure if we can make a meaningful progress without one.
